### PR TITLE
Issue 467 by Pjohans: make string strict cql for well 3.2

### PR DESCRIPTION
**Pull request by pjohans**

### CODE REVIEW

File: ting_search.module [hunk 1]
```diff
@@ -967,36 +967,50 @@ function ting_search_submit($form, &$form_state) {
 }
 
 /**
- * Attempt to quote reserved words in a search query.
- *
- * As proper quoting would require a full CQL parser, we cheat and
- * just work on the part we know is the free text part.
- *
- * Also, we don't mess with uppercase reserved words.
+ * Make sure string is strict cql
```
Comment (cableman) [inline ting_search.module]: Missing parameter comments in function documentation.

File: ting_search.module [hunk 2]
```diff
@@ -967,36 +967,50 @@ function ting_search_submit($form, &$form_state) {
...
-  if (preg_match('/^(.*?)(AND \(.*|$)/', $string, $rx)) {
-    $keys = $rx[1];
-    $new_keys = preg_replace_callback('/(?:(".*?(?<!\\\)")|\b(and|or|not|prox)\b)/i', '_ting_search_quote_callback', $keys);
-    $string = preg_replace('/^' . preg_quote($keys, '/') . '/', $new_keys, $string);
+  return ting_search_string_to_cql($string);
+}
+
+/**
+ * Method to convert a string to strict cql (Basically by adding quotes when needed)
+ * @param $string
```
Comment (cableman) [inline ting_search.module]: Missing parameter comments in function documentation.

File: ting_search.module [hunk 3]
```diff
@@ -967,36 +967,50 @@ function ting_search_submit($form, &$form_state) {
...
+  if (isset($string) && !_ting_search_string_is_cql($string)) {
+    $string = _ting_search_convert_string_to_cql($string);
   }
   return $string;
 }
 
 /**
- *
+ * Add quotes to string and escapes existing quotes
+ * @param $string
```
Comment (cableman) [inline ting_search.module]: Missing parameter comments in function documentation.

File: ting_search.module [hunk 4]
```diff
@@ -967,36 +967,50 @@ function ting_search_submit($form, &$form_state) {
...
+  $string = preg_replace('/\s+/', ' ', $string);
+  // Explode string by whitespace.
+  $expressions = explode(' ', $string);
+  // Join expressions with 'and' operator.
+  return implode(' and ', $expressions);
+}
+
+/**
+ * Tests if a string is cql
+ * @param $string
```
Comment (cableman) [inline ting_search.module]: Missing parameter comments in function documentation.

File: ting_search.module [hunk 5]
```diff
@@ -967,36 +967,56 @@ function ting_search_submit($form, &$form_state) {
...
+/**
+ * Method to convert a string to strict cql (Basically by adding quotes when needed)
+ * @param string $string
+ *   The search query
+ * @return string
+ *   cql compatible string
+ */
+function ting_search_string_to_cql($string) {
+  $string = trim($string);
+  if (isset($string) && !_ting_search_string_is_cql($string)) {
```
Comment (kasperg) [inline ting_search.module]: Why is `isset()` necessary here? I would think that `$string` would always be set at this point.

`trim()` always return a string anyway.

File: ting_search.module [hunk 6]
```diff
@@ -967,36 +967,56 @@ function ting_search_submit($form, &$form_state) {
...
-    return $matches[2];
+function _ting_search_convert_string_to_cql($string) {
+  // Escape quotes.
+  $string = str_replace('"', '\"', $string);
+  // Remove multiple whitespaces.
+  $string = preg_replace('/\s+/', ' ', $string);
+  // Explode string by whitespace.
+  $expressions = explode(' ', $string);
+  // Join expressions with 'and' operator.
+  return implode(' and ', $expressions);
```
Comment (kasperg) [inline ting_search.module]: I am no expert in CQL but this conversion seems odd to me.

As I understand it if i search for `"Harry Potter" fire` then that is converted to `\"Harry and Potter\" and fire'. That does not seem right?

Apart from the handling of quotes using `and` effectively narrows the search result. I would expect it to widen it and let the ranking move the best matches to the top of the result set.

File: ting_search.module [hunk 7]
```diff
@@ -967,36 +967,56 @@ function ting_search_submit($form, &$form_state) {
...
+ * @return bool|int
+ *   Whether the string is valid cql(TRUE) or not(FALSE)
+ */
+function _ting_search_string_is_cql($string) {
+  if(strpos($string, '"') !== false && (!preg_match('@ and | or | not |=@', $string) && !(substr($string, 0, 1) === '"' && substr($string, -1) === '"'))){
+    return false;
   }
-  // We have a quote. Just return it.
-  return $matches[1];
+  return preg_match('@"|\(|\)|=@', $string);
```
Comment (kasperg) [inline ting_search.module]: I would appreciate an explanation of why each of these tests determine if a string is CQL or not.

File: ting_search.module (hunk 6 above)
Comment (vibjerg) [inline ting_search.module]: It is a matter of guessing wether or not the user writes cql or not. In the case you describes, if the user tries to write cql, it would be invalid and should there for fail. If the user tries to find a title containing quotes, we will need to escape the quotes, and that is the first part of what is happening. 

When you search on a phrase, you can either put quotes around the phrase or 'and' between each word - the result is the same. 

in short :
\"Harry and Potter\" and fire == "\"Harry Potter\" fire"
"Harry and Potter" and fire  == invalid cql

Comment (kasperg) [inline ting_search.module]: In my opinion end users (citizens) do not write CQL. Their search string might be valid CQL but that is probably by accident. The reference point here is Google. Expecting them to learn CQL raises the barrier for searching in DDB CMS unreasonably in my opinion. If OpenSearch requires valid CQL search strings then it is our job to take the user search string and try our best to convert it to the corresponding CQL string based on their intent.

> When you search on a phrase, you can either put quotes around the phrase or 'and' between each word - the result is the same.

There is a significant difference between searching for the phrase `Harry Potter` and the word `fire` and the words `"Harry`, `Potter"` and `fire`.

`"Harry Potter" fire` is a valid search string in current installations and [the results](https://koegebib.dk/search/ting/%22harry%20potter%22%20fire) seem right to me. Are you telling me that after this change this search string would be converted to a CQL string for the words `"Harry`, `Potter"` and `fire` ?

**Comment by pjohans** [comment]
This is the first time i've been involved in this discussion. As i see it (in agreement with kasper) we should not expect the end user to write cql. but that is an issue to be further elaborated on.

Regarding the ["harry potter" fire] search string  - that should be converted to [\"harry potter\" and fire] to be valid and correct. As it is now it becomes [\"harry and potter\" and fire] and that is an error. I will take out the [and] after lunch

**Comment by vibjerg** [comment]
I Agree with Kasper. Citizens do NOT write CQL. But someone has required that it should be possible to write CQL. Therefore we need to guess if it cql or not. 

We characterise it as cql if query contains "()" or/and "=" with some specific combination of \", and, or, not. If not it is NOT CQL. 

The query ["harry potter" fire], we do not characterise as cql, but it is invalid cql and therefore we cannot send it directly to opensearch, so we have to transform it to valid cql. There easiest way to do that is by putting and between the words. 

"harry potter fire"
"harry potter" and fire
harry and potter and fire
\"harry and \"potter and fire. 

Will effectively give the same result. 

I do think though, that it would be better to remove the quotes, than escaping them, because they are normalized out by solr.

**Comment by cableman** [comment]
The request write CQL in search's comes form the library staff, which also uses the homepage as thier search interface and they can make some very accurate search with CQL.

**Comment by kasperg** [comment]
I have no problem with allowing the library staff to write CQL as long as we have proper solutions for:

1. Separating non-CQL queries from CQL ones
2. Converting non-CQL queries to CQL

Regarding 1. we have a working solution that I would like to see explained a bit more. 

If this matching does not work in the field we will have to look at other options down the line. Since this is a feature aimed specifically at librarians I think it is OK to add additional requirements - e.g. all CQL queries must be in the format "CQL:query, [query]" or the like.

I am questioning whether the solution to 2 works as expected. @pjohans will update the code to deal with the example discussed properly. That sounds like a good start!

**Comment by vibjerg** [comment]
The new solution seems to work as expected and is merged
